File: app/src/main/java/com/elementary/tasks/core/contacts/CallsAsync.java
package com.elementary.tasks.core.contacts;

import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.AsyncTask;
import android.provider.CallLog;

import com.elementary.tasks.core.utils.Contacts;

import java.util.ArrayList;
import java.util.List;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class CallsAsync extends AsyncTask<Void, Void, Void> {

    private List<CallsItem> mList;
    private Context mContext;
    private CallsLogListener mListener;

    public CallsAsync(Context context, CallsLogListener listener) {
        this.mContext = context;
        this.mListener = listener;
    }

    @Override
    protected Void doInBackground(Void... params) {
        Cursor c = mContext.getContentResolver().query(CallLog.Calls.CONTENT_URI, null, null, null, null);
        mList = new ArrayList<>();
        mList.clear();
        if (c != null) {
            int number = c.getColumnIndex(CallLog.Calls.NUMBER);
            int type = c.getColumnIndex(CallLog.Calls.TYPE);
            int date = c.getColumnIndex(CallLog.Calls.DATE);
            int nameIndex = c.getColumnIndex(CallLog.Calls.CACHED_NAME);
            while (c.moveToNext()) {
                String phoneNumber = c.getString(number);
                String callType = c.getString(type);
                String callDate = c.getString(date);
                String name = c.getString(nameIndex);
                int id = Contacts.getIdFromNumber(phoneNumber, mContext);
                String photo = null;
                if (id != 0) {
                    Uri uri = Contacts.getPhoto(id);
                    if (uri != null) {
                        photo = uri.toString();
                    }
                }

                CallsItem data = new CallsItem(name, phoneNumber, photo, Long.valueOf(callDate), id, Integer.parseInt(callType));
                int pos = getPosition(data.getDate());
                if (pos == -1) {
                    mList.add(data);
                } else {
                    mList.add(pos, data);
                }
            }
            c.close();
        }
        return null;
    }

    private int getPosition(long date) {
        if (mList.size() == 0) {
            return 0;
        }
        int position = -1;
        for (CallsItem data : mList) {
            if (date > data.getDate()) {
                position = mList.indexOf(data);
                break;
            }
        }
        return position;
    }

    @Override
    protected void onPostExecute(Void aVoid) {
        super.onPostExecute(aVoid);
        if (mListener != null) {
            mListener.onLoaded(mList);
        }
    }
}

File: app/src/main/java/com/elementary/tasks/core/contacts/calls/CallsAsync.kt
package com.elementary.tasks.core.contacts.calls

import android.content.Context
import android.os.AsyncTask
import android.provider.CallLog
import com.elementary.tasks.core.utils.Contacts
import com.elementary.tasks.core.utils.Permissions

/**
 * Copyright 2016 Nazar Suhovich
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
class CallsAsync(private val mContext: Context, private val mListener: ((List<CallsItem>) -> Unit)?) : AsyncTask<Void, Void, Void>() {

    private val mList: MutableList<CallsItem> = mutableListOf()

    override fun doInBackground(vararg params: Void): Void? {
        if (Permissions.checkPermission(mContext, Permissions.READ_CALLS)) {
            val c = mContext.contentResolver.query(CallLog.Calls.CONTENT_URI, null, null, null, null)
            mList.clear()
            if (c != null) {
                val number = c.getColumnIndex(CallLog.Calls.NUMBER)
                val type = c.getColumnIndex(CallLog.Calls.TYPE)
                val date = c.getColumnIndex(CallLog.Calls.DATE)
                val nameIndex = c.getColumnIndex(CallLog.Calls.CACHED_NAME)
                while (c.moveToNext()) {
                    val phoneNumber = c.getString(number)
                    val callType = c.getString(type)
                    val callDate = c.getString(date)
                    val name = c.getString(nameIndex)
                    val id = Contacts.getIdFromNumber(phoneNumber, mContext)
                    var photo: String? = null
                    if (id != 0) {
                        val uri = Contacts.getPhoto(id.toLong())
                        if (uri != null) {
                            photo = uri.toString()
                        }
                    }

                    val data = CallsItem(name, phoneNumber, photo, java.lang.Long.valueOf(callDate), id, Integer.parseInt(callType))
                    val pos = getPosition(data.date)
                    if (pos == -1) {
                        mList.add(data)
                    } else {
                        mList.add(pos, data)
                    }
                }
                c.close()
            }
        }
        return null
    }

    private fun getPosition(date: Long): Int {
        if (mList.size == 0) {
            return 0
        }
        var position = -1
        for (data in mList) {
            if (date > data.date) {
                position = mList.indexOf(data)
                break
            }
        }
        return position
    }

    override fun onPostExecute(aVoid: Void) {
        super.onPostExecute(aVoid)
        mListener?.invoke(mList.toList())
    }
}
